src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/ExceptionHandling.wasm.cs

```diff
@@ -117,6 +120,48 @@ private static void HandleExceptionWasmFault(void* pShadowFrame, DispatchData* p
             DispatchContinueSearch(exception, pDispatchData);
         }
 
+        // This handler is called by codegen for exceptions that escape from RPI methods (i. e. unhandled exceptions).
+        //
+        private static void HandleUnhandledException(object exception)
+        {
+            // We have to duplicate the logic because the helpers are runtime exports and need a calling convention conversion.
+            // Also, we cannot use code addresses to get helpers.
+            IntPtr pOnUnhandledExceptionFunction =
+                (IntPtr)InternalCalls.RhpGetClasslibFunctionFromEEType(exception.GetMethodTable(), ClassLibFunctionId.OnUnhandledException);
+
+            if (pOnUnhandledExceptionFunction != IntPtr.Zero)
+            {
+                try
+                {
+                    InternalCalls.RhpRawCalli_VO(pOnUnhandledExceptionFunction, exception);
+                }
+                catch when (true)
+                {
+                    // disallow all exceptions leaking out of callbacks
+                }
+            }
+
+            IntPtr pFailFastFunction =
+                (IntPtr)InternalCalls.RhpGetClasslibFunctionFromEEType(exception.GetMethodTable(), ClassLibFunctionId.FailFast);
+
+            if (pFailFastFunction == IntPtr.Zero)
+            {
+                FallbackFailFast(RhFailFastReason.UnhandledException, exception);
+            }
+
+            try
+            {
+                InternalCalls.RhpRawCalli_ViOII(pFailFastFunction, (int)RhFailFastReason.UnhandledException, exception, 0, 0);
+            }
+            catch when (true)
```

Comment:
Is the `when (true)` a style guide?

Reply:
No - it's required for correctness in the presense of multiple type universes. A catch-all `catch` is emitted as `catch (object)` by Roslyn, which would "burn" one `System.Object` into the EH table, and that `System.Object` may not be the same `System.Object` that the thrower observes.

(This capability is currently not used as the CoreLib and runtime are merged, but we keep it working)

Reply:
Thanks!